File: Data Structures & Algorithms/course-schedule-ii/submission-2.java
class Solution {
    private int topIdx;

    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<List<Integer>> adj = new ArrayList<>();

        for (int i = 0; i < numCourses; i++) adj.add(new ArrayList<>());

        for (int[] p : prerequisites) adj.get(p[0]).add(p[1]);

        int[] visited = new int[numCourses]; // 0 = unvisited, 1 = visiting, 2 = visited
        int[] res = new int[numCourses];
        topIdx = 0;

        for (int i = 0; i < numCourses; i++) {
            if (!dfs(i, adj, res, visited)) return new int[0];
        }

        return res;
    }

    private boolean dfs(int i, List<List<Integer>> adj, int[] res, int[] visited) {
        if (visited[i] == 1) return false;
        if (visited[i] == 2) return true;

        visited[i] = 1;

        List<Integer> children = adj.get(i);
        
        for (int child : children) {
            if (!dfs(child, adj, res, visited)) return false;
        }

        res[topIdx++] = i;

        visited[i] = 2;

        return true;
    }
}

File: Data Structures & Algorithms/interleaving-string/submission-2.py
class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        n1, n2 = len(s1), len(s2)
        if n1 + n2 != len(s3):
            return False
        
        if len(s3) == 0:
            return True
        
        dp = {}

        def dfs(i1, i2):
            if i1 == n1 and i2 == n2:
                return True

            if (i1, i2) in dp:
                return dp[(i1, i2)]
            
            i3 = i1 + i2
            match = False

            if i1 < n1 and s1[i1] == s3[i3]:
                match = match or dfs(i1 + 1, i2)
            
            if i2 < n2 and s2[i2] == s3[i3]:
                match = match or dfs(i1, i2 + 1)
            
            dp[(i1, i2)] = match
            return match
            
        return dfs(0, 0)